### ISSUE TITLE
Service consumes more than what is limited

### ISSUE THREAD

**Issue by cristhianbicca** [moby/moby#33526]
am using the version:

```
Client:
 Version: 17.04.0-ce
 API version: 1.28
 Go version: go1.7.5
 Git commit: 4845c56
 Built: Mon Apr 3 18:07:42 2017
 OS / Arch: linux / amd64

Server:
 Version: 17.04.0-ce
 API version: 1.28 (minimum version 1.12)
 Go version: go1.7.5
 Git commit: 4845c56
 Built: Mon Apr 3 18:07:42 2017
 OS / Arch: linux / amd64
 Experimental: true
```

The problem is the following I accomplished the limitation of the service with the parameter `--limit-memory 256000000`. When giving the docker stats in the container, it shows me the limitation of:

```
CONTAINER           CPU %               MEM USAGE / LIMIT     MEM %               NET I/O             BLOCK I/O           PIDS
2dece58afef9        0.00%               243.8MiB / 244.1MiB   99.85%              21MB / 11.5MB       147MB / 478MB       199
```

```
systemd-cgls
Working directory /sys/fs/cgroup/memory/docker/2dece58afef92c66cd21de9d1f49603f53764fdf6221cd0d6906d70fd217694f:
├─41594 /bin/bash
├─48291 /bin/bash /run.sh
├─48492 npm
├─49450 sh -c node app
├─49451 node app
├─49459 /usr/local/bin/node /usr/src/app/app
├─49465 /usr/local/bin/node /usr/src/app/app
├─49466 /usr/local/bin/node /usr/src/app/app
├─49478 /usr/local/bin/node /usr/src/app/app
├─49485 /usr/local/bin/node /usr/src/app/app
├─49496 /usr/local/bin/node /usr/src/app/app
├─49497 /usr/local/bin/node /usr/src/app/app
├─49509 /usr/local/bin/node /usr/src/app/app
├─49511 /usr/local/bin/node /usr/src/app/app
├─49513 /usr/local/bin/node /usr/src/app/app
├─49526 /usr/local/bin/node /usr/src/app/app
├─49531 /usr/local/bin/node /usr/src/app/app
├─49541 /usr/local/bin/node /usr/src/app/app
├─49567 /usr/local/bin/node /usr/src/app/app
├─49569 /usr/local/bin/node /usr/src/app/app
├─49592 /usr/local/bin/node /usr/src/app/app
├─49604 /usr/local/bin/node /usr/src/app/app
├─49606 /usr/local/bin/node /usr/src/app/app
├─49611 /usr/local/bin/node /usr/src/app/app
├─49659 /usr/local/bin/node /usr/src/app/app
├─50165 /usr/local/bin/node /usr/src/app/app
├─50272 /usr/local/bin/node /usr/src/app/app
├─50732 /usr/local/bin/node /usr/src/app/app
├─50865 /usr/local/bin/node /usr/src/app/app
├─51411 /usr/local/bin/node /usr/src/app/app
├─51412 /usr/local/bin/node /usr/src/app/app
├─51437 /usr/local/bin/node /usr/src/app/app
├─51440 /usr/local/bin/node /usr/src/app/app
├─51461 /usr/local/bin/node /usr/src/app/app
└─51472 /usr/local/bin/node /usr/src/app/app
```

But adding the memory used by all the processes in my container, the sum of them is around 400Mb, does anyone know if it is normal behavior and why?

Thank you

**Comment by thaJeztah** [comment]
What distro / setup are you running on? Can you provide the output of `docker info`? Is memory limit supported on your machine? (you can use the `check-config` script to check if it's enabled https://github.com/moby/moby/blob/master/contrib/check-config.sh)

**Comment by cristhianbicca** [comment]
@thaJeztah  distro is `16.04.1 LTS (GNU/Linux 4.4.0-78-generic x86_64)`

output `docker info`

root@server:/opt# docker info
Containers: 173
 Running: 43
 Paused: 0
 Stopped: 130
Images: 10
Server Version: 17.04.0-ce
Storage Driver: devicemapper
 Pool Name: data-thinpool
 Pool Blocksize: 524.3kB
 Base Device Size: 10.74GB
 Backing Filesystem: xfs
 Data file:
 Metadata file:
 Data Space Used: 41.59GB
 Data Space Total: 214.7GB
 Data Space Available: 173.2GB
 Metadata Space Used: 41.31MB
 Metadata Space Total: 16.98GB
 Metadata Space Available: 16.94GB
 Thin Pool Minimum Free Space: 21.47GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.110 (2015-10-30)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: active
 NodeID: 
 Is Manager: false
 Node Address: 192.192.0.6
 Manager Addresses:
  192.192.0.2:2377
  192.192.0.3:2377
  192.192.0.4:2377
Runtimes: runc
Default Runtime: runc
Init Binary:
containerd version: 422e31ce907fd9c3833a38d7b8fdd023e5a76e73
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-72-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 30
Total Memory: 49.13GiB
Name: 
ID: 
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support

output `check-config`

`info: reading kernel config from /boot/config-4.4.0-78-generic ...

Generally Necessary:
- cgroup hierarchy: properly mounted [/sys/fs/cgroup]
- apparmor: enabled and tools installed
- CONFIG_NAMESPACES: enabled
- CONFIG_NET_NS: enabled
- CONFIG_PID_NS: enabled
- CONFIG_IPC_NS: enabled
- CONFIG_UTS_NS: enabled
- CONFIG_CGROUPS: enabled
- CONFIG_CGROUP_CPUACCT: enabled
- CONFIG_CGROUP_DEVICE: enabled
- CONFIG_CGROUP_FREEZER: enabled
- CONFIG_CGROUP_SCHED: enabled
- CONFIG_CPUSETS: enabled
- CONFIG_MEMCG: enabled
- CONFIG_KEYS: enabled
- CONFIG_VETH: enabled (as module)
- CONFIG_BRIDGE: enabled (as module)
- CONFIG_BRIDGE_NETFILTER: enabled (as module)
- CONFIG_NF_NAT_IPV4: enabled (as module)
- CONFIG_IP_NF_FILTER: enabled (as module)
- CONFIG_IP_NF_TARGET_MASQUERADE: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_ADDRTYPE: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_CONNTRACK: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_IPVS: enabled (as module)
- CONFIG_IP_NF_NAT: enabled (as module)
- CONFIG_NF_NAT: enabled (as module)
- CONFIG_NF_NAT_NEEDED: enabled
- CONFIG_POSIX_MQUEUE: enabled
- CONFIG_DEVPTS_MULTIPLE_INSTANCES: enabled

Optional Features:
- CONFIG_USER_NS: enabled
- CONFIG_SECCOMP: enabled
- CONFIG_CGROUP_PIDS: enabled
- CONFIG_MEMCG_SWAP: enabled
- CONFIG_MEMCG_SWAP_ENABLED: missing
    (cgroup swap accounting is currently not enabled, you can enable it by setting boot option "swapaccount=1")
- CONFIG_LEGACY_VSYSCALL_EMULATE: enabled
- CONFIG_MEMCG_KMEM: enabled
- CONFIG_BLK_CGROUP: enabled
- CONFIG_BLK_DEV_THROTTLING: enabled
- CONFIG_IOSCHED_CFQ: enabled
- CONFIG_CFQ_GROUP_IOSCHED: enabled
- CONFIG_CGROUP_PERF: enabled
- CONFIG_CGROUP_HUGETLB: enabled
- CONFIG_NET_CLS_CGROUP: enabled (as module)
- CONFIG_CGROUP_NET_PRIO: enabled
- CONFIG_CFS_BANDWIDTH: enabled
- CONFIG_FAIR_GROUP_SCHED: enabled
- CONFIG_RT_GROUP_SCHED: missing
- CONFIG_IP_VS: enabled (as module)
- CONFIG_IP_VS_NFCT: enabled
- CONFIG_IP_VS_RR: enabled (as module)
- CONFIG_EXT4_FS: enabled
- CONFIG_EXT4_FS_POSIX_ACL: enabled
- CONFIG_EXT4_FS_SECURITY: enabled
- Network Drivers:
  - "overlay":
    - CONFIG_VXLAN: enabled (as module)
      Optional (for encrypted networks):
      - CONFIG_CRYPTO: enabled
      - CONFIG_CRYPTO_AEAD: enabled (as module)
      - CONFIG_CRYPTO_GCM: enabled (as module)
      - CONFIG_CRYPTO_SEQIV: enabled (as module)
      - CONFIG_CRYPTO_GHASH: enabled (as module)
      - CONFIG_XFRM: enabled
      - CONFIG_XFRM_USER: enabled (as module)
      - CONFIG_XFRM_ALGO: enabled (as module)
      - CONFIG_INET_ESP: enabled (as module)
      - CONFIG_INET_XFRM_MODE_TRANSPORT: enabled (as module)
  - "ipvlan":
    - CONFIG_IPVLAN: enabled (as module)
  - "macvlan":
    - CONFIG_MACVLAN: enabled (as module)
    - CONFIG_DUMMY: enabled (as module)
  - "ftp,tftp client in container":
    - CONFIG_NF_NAT_FTP: enabled (as module)
    - CONFIG_NF_CONNTRACK_FTP: enabled (as module)
    - CONFIG_NF_NAT_TFTP: enabled (as module)
    - CONFIG_NF_CONNTRACK_TFTP: enabled (as module)
- Storage Drivers:
  - "aufs":
    - CONFIG_AUFS_FS: enabled (as module)
  - "btrfs":
    - CONFIG_BTRFS_FS: enabled (as module)
    - CONFIG_BTRFS_FS_POSIX_ACL: enabled
  - "devicemapper":
    - CONFIG_BLK_DEV_DM: enabled
    - CONFIG_DM_THIN_PROVISIONING: enabled (as module)
  - "overlay":
    - CONFIG_OVERLAY_FS: enabled (as module)
  - "zfs":
    - /dev/zfs: missing
    - zfs command: missing
    - zpool command: missing

Limits:
- /proc/sys/kernel/keys/root_maxkeys: 1000000
`

**Comment by cristhianbicca** [comment]
@thaJeztah

**Comment by thaJeztah** [comment]
Sorry for the long delay, this slipped under my radar.

Actually, this may be related to the fix applied in https://github.com/moby/moby/pull/32777 (https://github.com/docker/cli/pull/80), which is included in docker 17.06 and up; are you still seeing this on recent versions of docker?

**Comment by thaJeztah** [comment]
closing, as this went stale
